support-frontend/app/services/PayPalNvpService.scala

```diff
@@ -90,14 +92,51 @@ class PayPalNvpService(apiConfig: PayPalConfig, wsClient: WSClient) extends Touc
       "RETURNURL" -> returnUrl,
       "CANCELURL" -> cancelUrl,
       "BILLINGTYPE" -> "MerchantInitiatedBilling",
-      "NOSHIPPING" -> "1"
+      "NOSHIPPING" -> noShipping
     )
 
     nvpRequest(paymentParams).map { resp =>
       retrieveNVPParam(resp, "TOKEN")
     }
   }
 
+  // Sets up a payment by contacting PayPal and also fetches the users details.
+  def createAgreementAndRetrieveUser(token: Token): Future[Option[PayPalCheckoutDetails]] =
+    createBillingAgreement(token).flatMap(
+      maybeBaid =>
+      retrieveUserInformation(token).map(maybeUserDetails =>
+        maybeBaid match {
+          case Some(baid) => Some(PayPalCheckoutDetails(baid, maybeUserDetails))
+          case _ => None
+        }
+      )
+    )
+
+  def retrieveUserInformation(token: Token): Future[Option[PayPalUserDetails]] = {
+    val paymentParams = Map(
+      "METHOD" -> "GetExpressCheckoutDetails",
+      "TOKEN" -> token.token
+    )
+
+    nvpRequest(paymentParams).map { resp =>
+      val userDetails = for {
+        firstName <- retrieveNVPParam(resp, "FIRSTNAME")
+        lastName <- retrieveNVPParam(resp, "LASTNAME")
+        email <- retrieveNVPParam(resp, "EMAIL")
+        shipToStreet <- retrieveNVPParam(resp, "PAYMENTREQUEST_0_SHIPTOSTREET")
+        shipToCity <- retrieveNVPParam(resp, "PAYMENTREQUEST_0_SHIPTOCITY")
+        shipToState = retrieveNVPParam(resp, "PAYMENTREQUEST_0_SHIPTOSTATE") // State/County may not be present
+        shipToZip <- retrieveNVPParam(resp, "PAYMENTREQUEST_0_SHIPTOZIP")
+        shipToCountryCode <- retrieveNVPParam(resp, "PAYMENTREQUEST_0_SHIPTOCOUNTRYCODE")
+      } yield PayPalUserDetails(
+        firstName, lastName, email, shipToStreet, shipToCity, shipToState, shipToZip, shipToCountryCode
+      )
+      if (userDetails.isEmpty)
+        SafeLogger.error(scrub"Unable to retrieve user details from PayPal with token $token, nvp response was $resp")
```

Comment:
I think you said that for contributions this will be expected to be None?  So we will end up with a lot of worrying log spam if we use it there in future.
I would be tempted to return an Either and then log it in the caller at the point we deem that it's an actual error.

Reply:
Yeah this is a bit difficult to do unfortunately because all the different PayPal call sites call the same function and we don't know on the server when it's ok for there not to be user and when it's required. Again, this is a test and we're not expecting to leave it in place so I'm just going to remove this logging for now. If we include something like this in future we'll need to have a look at refactoring.